### PR TITLE
Fix adv_parser for Switchbot Lock Pro

### PR DESCRIPTION
**Pull request by szclsya**
Turns out the lock pro indeed has a different passive mode data format (duh), and we will need a separate parser for it.

Currently only these fields are tested to work:
+ battery
+ calibration
+ status
+ door_open
+ unclosed_alarm
+ unlocked_alarm
+ auto_lock_pause

And these are taken directly from the original lock, thus needs more testing on appropriate hardware:
+ update_from_secondary_lock
+ double_lock_mode
+ night_latch

Put it in `adv_parsers/lock.py` since code for lock pro active mode also lives side by side with the original lock at `devices/lock.py`.

### CODE REVIEW

**Comment by Michal4K** [comment]
https://github.com/Danielhiversen/pySwitchbot/issues/233#issuecomment-2211983234
It looks night_latch work with this changes too as @jsprotky post his testing...

**Comment by szclsya** [comment]
Yup, I'll remove the "haven't been tested" comment real quick.

**Comment by szclsya** [comment]
> Let me know how I can help with checking `update_from_secondary_lock` and `double_lock_mode`. I see `update_from_secondary_lock` changes when unlatching the lock for some statuses but I don't know where it is used at all. Don't have a clue what `double_lock_mode` is tbh.

According to switchbot documentation [switchbot's doc](https://support.switch-bot.com/hc/en-us/articles/8095324305047-How-to-Set-up-Dual-Lock-for-SwitchBot-Lock) it's used when you group two locks together.

**Comment by szclsya** [comment]
> Can not help here. Only have 1 lock.

Yeah me neither lol. 

I guess we can just return False for both of them for now? As these features are kinda niche and I believe people really want the other part to work. We can always reintroduce these fields later, when proper data shows up.

**Comment by szclsya** [comment]
@jsprotky That's because the old `update_from_secondary_lock` bit is now used by lock status. Check the changelog, and you can see the lock status now uses 3-5 bits of the 7th byte in mfr_data, while the existing `update_from_secondary_lock` parser uses the 5th bit of the 7th byte too, thus the confusion.

I was confused by it too, but since it seems that 3-5 bits do correctly parsed into lock status it should be just a rearrange.

**Comment by schummar** [comment]
According to the website, the Pro does not support Dual Lock (yet?): https://support.switch-bot.com/hc/en-us/articles/22917404813207-Can-Two-SwitchBot-Lock-Pro-Devices-Be-Set-up-for-Dual-Lock
I have two EU Lock Pros and there is indeed no option in the app.

I have found two issues with the current state, which might be solved by this PR, or maybe there is something else missing:
- When the lock is not locked, but not unlatched either, it's always displaying as "Unlocking" in HA.
- The lock is not displayed as having the unlock without unlatch option in HA. Apparently for that the device has to have the LockEntityFeature.OPEN flag set. (See e.g. https://github.com/home-assistant/frontend/blob/bef53aef576d2eb3982fbfa4bfcf146396697039/src/dialogs/more-info/controls/more-info-lock.ts#L115C9-L138C21)

I'm happy to assist with testing or might try a fix myself, if I have the time.

File: switchbot/adv_parsers/lock.py [hunk 1]
```diff
@@ -29,3 +29,28 @@ def process_wolock(data: bytes | None, mfr_data: bytes | None) -> dict[str, bool
...
+    data: bytes | None, mfr_data: bytes | None
+) -> dict[str, bool | int]:
+    _LOGGER.debug("mfr_data: %s", mfr_data.hex())
+    if data:
+        _LOGGER.debug("data: %s", data.hex())
+
+    res = {
+        "battery": data[2] & 0b01111111 if data else None,
+        "calibration": bool(mfr_data[7] & 0b10000000),
+        "status": LockStatus(int(mfr_data[7] & 0b00111000) / 8),
```
[the remark this answers is not in the document]
Comment (szclsya) [reply switchbot/adv_parsers/lock.py:44]: Ah it was some leftover of debugging to see if they changed how the lock status is encoded (thankfully they didn't). I'll change it to `>>3`.

**Comment by szclsya** [comment]
The new `manual` field seems to be either bugged or I misunderstood the flag, as it shows true even if the lock is locked by auto lock. Should we still keep it there?

**Comment by mwolter805** [comment]
@szclsya do you know if the integration can detect the state of both batteries when using the rechargeable [Dual Power Pack](https://us.switch-bot.com/products/switchbot-lock-pro?variant=47258771521769)? The app does report each separately. Please let me know if you need any info.
IMG_0458

**Comment by szclsya** [comment]
@mwolter805 As mentioned in the [issue](https://github.com/Danielhiversen/pySwitchbot/issues/233#issuecomment-2209833104) please provide passive scan data.

But I doubt this data is actually sent passively. It's more likely to be a response of a active mode command.

**Comment by ollo69** [comment]
> do you know if the integration can detect the state of both batteries when using the rechargeable [Dual Power Pack](https://us.switch-bot.com/products/switchbot-lock-pro?variant=47258771521769)? The app does report each separately. Please let me know if you need any info.

I already tried to search this info in the passive data but I didn't find any related information as I reported [here](https://github.com/Danielhiversen/pySwitchbot/issues/233#issuecomment-2211182237). Obviously would be great if I'm wrong 😉

**Comment by mwolter805** [comment]
Please forgive my ignorance, can I create and run that python script in the home assistant container?

**Comment by szclsya** [comment]
@mwolter805 Just as @ollo69 speculated the detailed battery info is indeed not sent in passive mode, so I'd assume there's a separate active mode command to check these information.

Anyhow it will be out of scope of this PR (since it's about paring passive mode data). Maybe open a separate issue or asking in the general Lock Pro issue?

**Review by bdraco — APPROVED** [review]

**Comment by bdraco** [comment]
Any more changes, or is this ready to merge?

**Comment by szclsya** [comment]
@bdraco Not sure if we should add the manual field (it's incorrect when locked by auto lock) but other than that we should be good.

**Comment by bdraco** [comment]
> @bdraco Not sure if we should add the manual field (it's incorrect when locked by auto lock) but other than that we should be good.

I'd keep it out for now unless its strictly needed for HA

**Comment by szclsya** [comment]
@bdraco All good now then.

**Review by bdraco — APPROVED** [review]

**Comment by bdraco** [comment]
https://github.com/Danielhiversen/pySwitchbot/compare/0.48.0...0.48.1

**Comment by bdraco** [comment]
This could also use a test with a few different advertisements so we know it won't break in the future.

**Comment by szclsya** [comment]
@bdraco True. Will add some in another PR when I have time.

**Comment by ollo69** [comment]
@szclsya,

if you don't have time I can create bump PR in HA for you, let me know.?

**Comment by szclsya** [comment]
@ollo69 Yeah sure, go ahead!

**Pull request merged**